packages/daimo-api/src/contract/ClogMatcher.ts

```diff
@@ -0,0 +1,309 @@
+          .where("tx_hash", "in", txList)
+          .where("log_name", "=", "FastFinish")
+          .execute()
+    );
+
+    // For each FastFinish event, load the corresponding Start event
+    for (const row of fastFinishResult) {
+      const startLog = await retryBackoff(
+        `StartFastCctp-logs-query-${from}-${to}`,
+        () =>
+          kdb
+            .selectFrom("index.daimo_fast_cctp")
+            .selectAll()
+            .where("chain_id", "=", row.from_chain_id)
+            .where("handoff_addr", "=", row.handoff_addr)
+            .where("nonce", "=", row.nonce)
+            .where("log_name", "=", "Start")
+            .executeTakeFirst()
+      );
+      if (!startLog) continue;
+
+      const fromTokenAddr = assertNotNull(startLog.from_token);
+      const fastCctpLog = {
+        blockHash: bytesToHex(row.block_hash, { size: 32 }),
+        blockNumber: BigInt(row.block_num),
+        transactionHash: bytesToHex(row.tx_hash, { size: 32 }),
+        transactionIndex: Number(row.tx_idx),
+        logIndex: Number(row.log_idx),
+        address: getAddress(bytesToHex(fromTokenAddr, { size: 20 })),
+        from: getAddress(
+          bytesToHex(assertNotNull(row.from_addr), { size: 20 })
+        ),
+        to: getAddress(bytesToHex(row.to_addr, { size: 20 })),
+        value: BigInt(row.to_amount),
+      };
+
+      const crossChainToken = this.tokenReg.getToken(
+        fastCctpLog.address,
+        Number(row.from_chain_id), // cross-chain (from chain)
+        true
+      );
+      if (crossChainToken == null) continue;
+
+      const crossChainTransfer: ForeignTokenTransfer = {
+        ...fastCctpLog,
+        foreignToken: crossChainToken,
+      };
+      this.txHashToTransfers.set(crossChainTransfer.transactionHash, [
+        ...(this.txHashToTransfers.get(crossChainTransfer.transactionHash) ||
+          []),
+        crossChainTransfer,
+      ]);
+    }
+
+    const elapsedMs = (Date.now() - startTime) | 0;
+    console.log(
+      `[SWAP] loaded ${fastFinishResult.length} FastCCTP cross-chain events in ${elapsedMs}ms`
+    );
+  }
+
+  /**
+   * For a multi-log outbound transfer (e.g. SwapClog, cross-chain transfer),
+   * attach the corresponding transfer info.
+   */
+  getCorrespondingTransfer(
+    from: Address,
+    to: Address,
+    transactionHash: string,
+    isInbound?: boolean
+  ) {
+    const transfers = this.txHashToTransfers.get(transactionHash);
+    if (!transfers || transfers.length === 0) return null;
+
+    // First check whether the transfer is cross-chain
+    if (to === daimoFastCctpAddress || isInbound) {
+      return this.getFastCctpTransfer(transfers);
+    }
+
+    // Then check for outbound swap
+    return this.getMatchingSwapTransfer(from, transfers);
+  }
+
+  /** For a cross-chain transfer, attach the corresponding transfer info. */
+  getFastCctpTransfer(
+    transfers: ForeignTokenTransfer[]
+  ): ForeignTokenTransfer | null {
```

Comment:
can simplify code below to:
`transfer.find((t) => t.foreignToken.chainId !== chainConfig.chainL2.id)`

...but what guarantees that'll be a fastcctp transfer?